ZFSin/zfs/module/zfs/zio.c

```diff
@@ -3605,7 +3605,7 @@ zio_vdev_io_done(zio_t *zio)
 
 	ops->vdev_op_io_done(zio);
 
-	if (unexpected_error && zio->io_waiter != NULL)
+	if (unexpected_error)
```

Comment:
Do we know why we have to make this change to the upstream code?

Reply:
When there is an EIO error coming from the disk ZFSin is currently not calling vdev_probe, which is actually responsible for putting the zpool into suspended state(by calling zio_suspend). Now in ZFSin we are repeatedly retrying to sync the transaction. The right behavior is to put the zpool into suspended state and the after fixing the disk error run 'zpool clear' command to remove suspend state.  So with **zio->io_waiter != NULL** check vdev_probe no getting called even though there is an error(I actually found this fix from openzfs).

Without this fix we will have a data miscompare issue, as my application is getting write success message even when zfs is getting EIO from the underlying disk(with sync=always and application is using NO_BUFFERING flag). With this fix we still have a miscompare, but its only for the first write after the disk error, before this fix all the writes after the disk error was wrongly acknowledged. The first write miscompare after disk error  is present in openzfs also,  I have also filed a bug in openzfs  **https://github.com/openzfs/zfs/issues/12391**.

Reply:
Ah perfect, thanks.